### ISSUE TITLE
what would something validate, and then crash on to_dict?

### ISSUE THREAD

**Issue by joefromct** [brunato/xmlschema#41]
Hi,

Is there any good reason to have a document validate, and then when `to_dict()` is called, receive a validation error such as `XMLSchemaChildrenValidationError`?

See below, thanks. 
```` python
import xmlschema as xs 

SCHEMA_FILE = "xsd.xsd"
SCHEMA = xs.XMLSchema(SCHEMA_FILE)
with open('../sample-input/xml.xml', mode="r") as fi:
    xml_file_content = fi.read()

print(SCHEMA.is_valid(xml_file_content))

# In [21]: True
`````
Then with the `to_dict()` I get the error:
```` python
d = SCHEMA.to_dict(xml_file_content)

#XMLSchemaChildrenValidationError: failed validating <Element '{http://domain.com/service/xmlobj}Value' at 0x7fc12115c6d8> with XsdElement(name='Boolean').
#
#Reason: The child n.1 of element 'Value' has a unexpected tag 'Integer'. Tag 'Boolean' expected.
# 
#Schema:
#
#    <xs:element xmlns:xs="http://www.w3.org/2001/XMLSchema" name="Boolean" type="my_value_boolean" />
#
#Instance:
#
#  <Value xmlns="http://domain.com/service/xmlobj">
#              <Integer>+00000000</Integer>
#  </Value>
#
````

Relevent portion of the xsd:
```` xml
<xs:complexType name="my_detail">
  <xs:sequence>
    <xs:element name="Name" type="xs:string"></xs:element>
    <xs:element name="Value">
      <xs:complexType>
        <xs:choice>
          <xs:element name="Boolean" type="xs:boolean" />
          <xs:element name="Integer" type="xs:int"     />
          <xs:element name="Decimal" type="xs:decimal" />
          <xs:element name="Code"    type="xs:string"  />
        </xs:choice>
      </xs:complexType>
    </xs:element>
  </xs:sequence>
</xs:complexType>

`````
a sample xml portion:
```` xml
...clip...
          <Detail>
            <Name>SomeNameValueThingy</Name>
            <Value>
              <Integer>+00000000</Integer>
            </Value>
          </Detail>
...clip...
`````

**Comment by AmatanHead** [comment]
So the problem seems to be in [groups.py:474](https://github.com/brunato/xmlschema/blob/master/xmlschema/validators/groups.py#L474). For each type in the choice element we try parsing that type. If parse was successful, we yield its result. If not, we try next element. If all types was tried without success, we yield an error.

The problem arise when `validation == 'strict'`. Instead of yielding an error, the parser throws it and [this `if` at line 475](https://github.com/brunato/xmlschema/blob/master/xmlschema/validators/groups.py#L475) cannot capture it.

A possible (and probably the most efficient) solution is to wrap subparsing into `try-except` statement.

**Comment by AmatanHead** [comment]
@brunato ping? This really slows my project.

**Comment by brunato** [comment]
I'm looking on that. 

Anyway the problem is a the generation of a validation error when the instance data is valid. With `validation == 'strict'` the decoder throws immediately in order to have a complete traceback (anyway it would throws into the top level function `decode()`).

The think the problem is on child_index value (1) that is not propagated (so the decoder retries and generates an error).  Still investigating why.

**Comment by wangnick-work** [comment]
Same issue here. Workaround for now is to `.decode(...,,validation='lax')`.

**Comment by brunato** [comment]
Yes, it was only a wrong parameter. The fix is available with the version 0.9.21.

Maybe the `iter_decode_children` method has become too complicated and I thought to rewrite this without generator functions (passing result_list as argument and index into a context object or a list...). Maybe for the next version.

Thanks
